File: src/util/entities/TeamMember.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2023 Spacebar and Spacebar Contributors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { Column, Entity, JoinColumn, ManyToOne, RelationId } from "typeorm";
import { BaseClass } from "./BaseClass";
import { User } from "./User";
import { TeamMemberRole, TeamMemberState } from "@spacebar/schemas";

@Entity({
	name: "team_members",
})
export class TeamMember extends BaseClass {
	@Column({ type: "int" })
	membership_state: TeamMemberState;

	@Column({ type: "simple-array" })
	permissions: string[];

	@Column()
	role: TeamMemberRole;

	@Column({ nullable: true })
	@RelationId((member: TeamMember) => member.team)
	team_id: string;

	@JoinColumn({ name: "team_id" })
	@ManyToOne(
		() => require("./Team").Team,
		(team: import("./Team").Team) => team.members,
		{
			onDelete: "CASCADE",
		},
	)
	team: import("./Team").Team;

	@Column({ nullable: true })
	@RelationId((member: TeamMember) => member.user)
	user_id: string;

	@JoinColumn({ name: "user_id" })
	@ManyToOne(() => User, {
		onDelete: "CASCADE",
	})
	user: User;
}

File: extra/admin-api/Models/Spacebar.Models.Db/Models/TeamMember.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Spacebar.Models.Db.Models;

[Table("team_members")]
public partial class TeamMember
{
    [Key]
    [Column("id")]
    public long Id { get; set; }

    [Column("membership_state")]
    public int MembershipState { get; set; }

    [Column("permissions")]
    public string Permissions { get; set; } = null!;

    [Column("team_id")]
    public long? TeamId { get; set; }

    [Column("user_id")]
    public long? UserId { get; set; }

    [Column("role", TypeName = "character varying")]
    public string Role { get; set; } = null!;

    [ForeignKey("TeamId")]
    [InverseProperty("TeamMembers")]
    public virtual Team? Team { get; set; }

    [ForeignKey("UserId")]
    [InverseProperty("TeamMembers")]
    public virtual User? User { get; set; }
}
